inbm/dispatcher-agent/dispatcher/sota/oras_util.py

```diff
@@ -0,0 +1,187 @@
+from ..packagemanager.irepo import IRepo
+from ..dispatcher_broker import DispatcherBroker
+from .constants import ORAS_TOKEN_PATH
+from .sota_error import SotaError
+
+logger = logging.getLogger(__name__)
+
+
+def oras_download(dispatcher_broker: DispatcherBroker, uri: CanonicalUri,
+             repo: IRepo, username: Optional[str], password: str, umask: int) -> None:
+    """Downloads files and places capsule file in path mentioned by manifest file.
+
+    @param dispatcher_broker: DispatcherBroker object used to communicate with other INBM services
+    @param uri: URI of the source location
+    @param repo: repository for holding the download
+    @param username: username to use for download
+    @param password: password to use for download
+    @param umask: file permission mask
+    @raises SotaError: any exception
+    """
+    dispatcher_broker.telemetry(f'Package to be fetched from {uri.value}')
+    dispatcher_broker.telemetry(
+        'Checking authenticity of package by checking signature and source')
+
+    if not isinstance(uri, CanonicalUri):
+        raise SotaError("Internal error: uri improperly passed to download function")
+
+    try:
+        source, registry_server, image, image_tag, repository_name, image_full_path, registry_manifest = \
+            parse_uri(uri)
+    except IndexError as err:
+        logger.error(f"IndexError occurs with uri {uri.value}: {err}")
+        raise SotaError(err)
+
+    logger.debug(f"source: {source}, "
+                 f"registry_server: {registry_server}, "
+                 f"image: {image}, "
+                 f"image_tag: {image_tag}, "
+                 f"repository_name: {repository_name}, "
+                 f"image_full_path: {image_full_path}, "
+                 f"registry_manifest: {registry_manifest}")
+
+    verify_source(source=source, dispatcher_broker=dispatcher_broker)
+    dispatcher_broker.telemetry('Source Verification check passed')
+
+    enough_space = is_enough_space_to_download(
+        registry_manifest, repo, password)
+
+    if not enough_space:
+        err_msg = " Insufficient free space available on " + repo.get_repo_path() + \
+                  " for " + str(uri.value)
+        raise SotaError(err_msg)
+
+    if password:
+        logger.debug("RS password provided.")
+    else:
+        err_msg = " No JWT token. Abort the update. "
+        raise SotaError(err_msg)
+
+    # Set password as environment variables for security reason.
+    env = dict(os.environ)  # make a copy of the environment
+    env["ORAS_PASSWORD"] = password
+
+    msg = f'Fetching software package from {image_full_path}'
+    dispatcher_broker.telemetry(msg)
+
+    # Call oras to pull the image. The password is the JWT token.
+    (out, err_run, code) = PseudoShellRunner().run(f"oras pull {image_full_path} -o {repo.get_repo_path()} "
+                                                   f"--password $ORAS_PASSWORD", env=env)
```

Comment:
don't put password on command line (can be read by `ps`)

Reply:
Let me try another way by `cat` the content of the access token file and passing it to oras command.

Reply:
Shell runner method updated. If password is provided, the password will be passed as stdin input when running the command.